Tail-modulo-cons manual chapter

gasche opened the pull request:
This is a proposal of documentation for tail-modulo-cons ( implementation submitted as #9760 ).

[Rendered output here](http://gallium.inria.fr/~scherer/tmp/tmc-manual/tailmodcons.html).

Feedback is welcome; this is an advanced feature (for expert users), but the manual must be readable (and reviewable) by people who don't know anything about the tail-mod-cons implementation.

For ease of reviewing, the manual addition is based on the current trunk, *not* the TMC branch. (So: if you want to play with the code, you need to checkout #9760 instead of the present PR.)

Note: there are some issues with the `caml_example` environment in this chapter, tracked and discussed in https://github.com/Octachron/ocaml/pull/5 . The present PR is in "Draft" mode as I replaced the `caml_example` environments with `verbatim`, manually including the output/warnings/errors. The commits are there to help debug the caml_example issues; I don't recommend looking at them individually to review the documentation.

gasche committed:
documentation chapter for [@tail_mod_cons]

wikku commented on manual/src/refman/extensions/tail_mod_cons.etex:
```suggestion
Again, the fix is to specialize "append_flatten" as well:
```

wikku commented on manual/src/refman/extensions/tail_mod_cons.etex:
```suggestion
intermediate representation of the OCaml compiler.)
```

wikku commented on manual/src/refman/extensions/tail_mod_cons.etex:
```suggestion
The call to "append_flatten", which syntactically appears in tail
position, gets transformed differently if the function has
a destination-passing-style version available
(i.e., is marked "[\@tail_mod_cons]"):
```

gasche replied:
I agree my version is awkward, but I don't like `i.e.` when addressing diverse users. I went for

> The call to "append_flatten", which syntactically appears in tail
position, gets transformed differently depending on whether the
function has a destination-passing-style version available, that is,
whether it is itself marked "[\@tail_mod_cons]":

wikku replied:
Looks fine to me.

wikku commented on manual/src/refman/extensions/tail_mod_cons.etex:
```suggestion
This explains the ``getting out of tail-mod-cons''
```

wikku commented on manual/src/refman/extensions/tail_mod_cons.etex:
```suggestion
Consider the call "Option.map foo x" for example: even if "foo" is
```

wikku commented on manual/src/refman/extensions/tail_mod_cons.etex:
```suggestion
position will get optimized. For example, if "List.map" is defined as
```

wikku commented on manual/src/refman/extensions/tail_mod_cons.etex:
```suggestion
indirectly, for example through function parameters.
```

gasche replied:
... I blame these on lack of sleep. Apologies.

wikku commented on manual/src/refman/extensions/tail_mod_cons.etex:
```suggestion
\paragraph{Non-exact calls to tupled functions} OCaml performs an
implicit optimization for ``tupled'' functions, which take a single
parameter that is a tuple: "let f (x, y, z) = ...". Direct calls to
these functions with a tuple literal argument (like "f (a, b, c)") will
call the ``tupled'' function by passing the parameters directly, instead
of building a tuple of them. Other calls, either indirect calls or calls
passing a more complex tuple value (like "let t = (a, b, c) in f t") are
compiled as ``inexact'' calls that go through a wrapper.
```

wikku commented on manual/src/refman/extensions/tail_mod_cons.etex:
```diff
@@ -0,0 +1,548 @@
+a result, this function consumes some amount of call-stack space on
+each recursive call. The stack usage of "map f li" is proportional to
+the length of "li". This is a correctness issue for large lists on
+operating systems with limited stack space -- the dreaded
```
```suggestion
operating systems with limited stack space --- the dreaded
```

dra27 replied:
This should be either `space -- the` or `space---the`, unless we already have uncapped em-dashes elsewhere.

gasche replied:
` -- ` is used often in the manual, and the ugly spaceless `---` is mostly only used in the flambda chapter -- probably authored by weird native English writers.

dra27 replied:
While I wouldn't call them weird (😁), I also prefer ` -- `

wikku commented on manual/src/refman/extensions/tail_mod_cons.etex:
```diff
@@ -0,0 +1,548 @@
+
+A well-known limitation of this implementation is that the recursive
+call, "map f xs", is not in \emph{tail} position. The runtime needs to
+remember to continue with "y :: _" after the call returns and, as
```
`y :: _` is syntax for continuations that wasn't explained anywhere, so maybe
```suggestion
remember to continue with "y :: r" after the call returns a value "r" and, as
```

wikku commented on manual/src/refman/extensions/tail_mod_cons.etex:
```suggestion
  calls'', if they occur in tail position or in tail-modulo-cons
```

gasche replied:
Is there something wrong with the "whether .. or" construct here? This is what I wanted to use and it seems clearer.

wikku replied:
It might be misread as "whether they occur in (tail position or tail-modulo-cons position) or not".
Most often (always?) in "whether A or B" the A and B exhaust all possibilities (often B is the negation or A, and often it might be written "or not" or even omitted). This is not the case here, as a call might be neither tail position nor tail-modulo-cons position.

dra27 replied:
One could further emphasise this by having "_even_ if they occur"

wikku replied:
Umm... "*only* if"?
```suggestion
  calls'', only if they occur in tail position or in tail-modulo-cons
```

dra27 replied:
Yes, that’s what I meant 😳

gasche replied:
I rephrased as follows:

> \item Calls from a tail-mod-cons function to another tail-mod-cons
  function declared in the same recursive-binding group are
  transformed into tail calls, as soon as they occur in tail position
  or tail-modulo-cons position in the source function.
\item Calls from a function \emph{not} marked tail-mod-cons to
  a tail-mod-cons function or, conversely, from a tail-mod-cons
  function to a non-tail-mod-cons function are transformed into
  \emph{non}-tail calls, even if they syntactically appear in tail
  position in the source program.

wikku commented on manual/src/refman/extensions/tail_mod_cons.etex:
```suggestion
\item The transformation is first-order: only direct calls to
  known tail-mod-cons functions become tail calls when in
```

wikku commented on manual/src/refman/extensions/tail_mod_cons.etex:
```suggestion
trees tend to be right-leaning (lots of "Let" in sequence,
```
```suggestion
trees tend to be unbalanced on the right (lots of "Let" in sequence,
```

dra27 commented on manual/src/refman/extensions/tail_mod_cons.etex:
```diff
@@ -0,0 +1,548 @@
+Note: this feature is considered experimental, and its interface may
+evolve, with user feedback, in the next few releases of the language.
+
+Consider the simple way of implementing the "List.map" function:
```
```suggestion
Consider the idiomatic way of implementing the "List.map" function:
```
??

gasche replied:
meh, "idiomatic" feels jargony to me, so I kept "simple" even if it is less precise.

dra27 replied:
"simple" to me has overtones (in this context, at least) of being naïve, which is unfortunate given that it's the actual implementation in the Standard Library!

Could it just be introduced that way instead? Perhaps `Consider the Standard Library's implementation of "List.map":`? I don't mind if "idiomatic" is too jargony (I don't agree FWIW - it's a quite standard English word), but "simple" doesn't feel appropriate 🤷

gasche replied:
I replaced "simple" by "natural", is it better?

(Personally I don't care that this implementation is in the standard library, and I would criticize it if made sense. But here the point is that this is the most obvious way to write `List.map`, not that it it is used in this or that library.)

dra27 replied:
"natural" works for me, thanks 🙂

ksromanov replied:
> meh, "idiomatic" feels jargony to me

But it is jargon of _our_ field! ;-) For me "idiomatic" means "natural for a given programming language". Of course just "natural" also looks good, but from point of view the decision is too cautious.

dra27 commented on manual/src/refman/extensions/tail_mod_cons.etex:
```diff
@@ -0,0 +1,548 @@
+
+Consider the simple way of implementing the "List.map" function:
+\begin{verbatim}
+let rec map f li =
```
The stdlib docs consistently (I think!) use `l` for lists. `li` reads oddly to me - could we have either `l` or `xs`/`ls`?

gasche replied:
I don't like `l` which looks too close to `|` or `1` for comfort on many fonts, but I forgot that `i` is a pronunciation minefield in English. Let's go with `l` for consistency, thanks.

dra27 commented on manual/src/refman/extensions/tail_mod_cons.etex:
```diff
@@ -0,0 +1,548 @@
+
+A well-known limitation of this implementation is that the recursive
+call, "map f xs", is not in \emph{tail} position. The runtime needs to
+remember to continue with "y :: _" after the call returns and, as
+a result, this function consumes some amount of call-stack space on
```
```suggestion
remember to continue with "y :: _" after the call returns and
this function consequently consumes some amount of call-stack space on
```
(I personally wouldn't use a subordinate clause here - the problem is a direct consequence)

dra27 commented on manual/src/refman/extensions/tail_mod_cons.etex:
```diff
@@ -0,0 +1,548 @@
+"Stack_overflow" exception.
+
+\begin{verbatim}
+# List.length (map Fun.id (List.init 1_000_000 (fun i -> i)));;
```
```suggestion
# List.length (map Fun.id (List.init 1_000_000 Fun.id));;
```

dra27 commented on manual/src/refman/extensions/tail_mod_cons.etex:
I'm finding it mildly confusing reading this that it was originally "tail modulo constructor" and is then shortened to "tail modulo cons" which makes it sound related to lists only. It might be "too late", but could it either be referred to in the text as "tail modulo constructor" consistently?

gasche replied:
`@tail_mod_cons` is the current attribute name (I originally used `@tmc` but people complained it was too obscure). I think that it is good to use this shorter name in the manual to make the connection clear, but I made it clear in my revision with an explicit mention of this shorter name.

dra27 commented on manual/src/refman/extensions/tail_mod_cons.etex:
```suggestion
# List.length (map Fun.id (List.init 1_000_000 Fun.id));;
```
(as before)

dra27 commented on manual/src/refman/extensions/tail_mod_cons.etex:
```suggestion
can be annotated to be tail position, helping establish stack-consumption
```
?

gasche replied:
Thanks, I went with "tail calls" instead.

dra27 commented on manual/src/refman/extensions/tail_mod_cons.etex:
```suggestion
one of these calls into a tail call. The compiler will not make a choice
```

dra27 commented on manual/src/refman/extensions/tail_mod_cons.etex:
```suggestion
implicitly, but ask the user to provide an explicit disambiguation.
```
? (though I like that you sort to not split your infinitives! 🙂)

ksromanov replied:
```tex
Instead of making an implicit choice, the compiler will ask the user to provide an explicit disambiguation.
```
?

gasche replied:
Thanks both for the suggestions. I went for

> The compiler will not make an implicit choice, but ask the user to provide an explicit disambiguation.

(I liked the symmetry between "implicitly" and "explicitly" we had before, and in this version inspired from @ksromanov we keep the symmetry between "implicit" and "explicit".)

dra27 commented on manual/src/refman/extensions/tail_mod_cons.etex:
```suggestion
Consider, for example, this type of syntactic expressions (assuming some
```
(or just drop the "for example" entirely)

dra27 commented on manual/src/refman/extensions/tail_mod_cons.etex:
```suggestion
recursive call that should be transformed to tail position:
```

dra27 commented on manual/src/refman/extensions/tail_mod_cons.etex:
```suggestion
Be aware that the resulting function is \emph{not} tail-recursive, the
```

dra27 commented on manual/src/refman/extensions/tail_mod_cons.etex:
I would write `na\"\ive`, but I'm not sure if that's a UK/US thing!

gasche replied:
I don't know what to do here so I will keep it as-is for now.

(Fun fact: the StackOverflow question ['naive' or 'naïve'](https://english.stackexchange.com/questions/12747/is-it-spelt-na%C3%AFve-or-naive) was closed as a duplicate of the question ['whereäs' as an alternative spelling of 'whereas'](https://english.stackexchange.com/questions/5603/where%c3%a4s-as-an-alternative-spelling-of-whereas).)

dra27 commented on manual/src/refman/extensions/tail_mod_cons.etex:
```suggestion
  function to a non-tail-mod-cons function are \emph{not} tail calls,
```

dra27 commented on manual/src/refman/extensions/tail_mod_cons.etex:
```suggestion
xss" will \emph{not} be tail calls. The correct fix here is to mark
```

dra27 commented on manual/src/refman/extensions/tail_mod_cons.etex:
```suggestion
Warning 51 [wrong-tailcall-expectation]: expected tailcall
```
??

ksromanov replied:
It is just a misprint here. Compilers code is correct:

```ocaml
  | Wrong_tailcall_expectation b ->
      Printf.sprintf "expected %s"
        (if b then "tailcall" else "non-tailcall")
```

https://github.com/ocaml/ocaml/blob/trunk/utils/warnings.ml#L951

dra27 commented on manual/src/refman/extensions/tail_mod_cons.etex:
```suggestion
In other cases, there is either no benefit in making the called function
tail-mod-cons, or it is not possible: for example, it is a function
```

dra27 commented on manual/src/refman/extensions/tail_mod_cons.etex:
This wasn't instantly clear to me, but I'm not sure of clearer wording: perhaps "function _argument_"? IIUC, the point is that an explicit call to TMC-annotated function `foo` can be transformed, but when passed as a value, `foo` "loses" the annotation? (the example below makes this clearer, so perhaps it doesn't matter too much)

ksromanov replied:
May be "to functions, known at compile-time"?

gasche replied:
In compilation terminology this is called a "direct call", so I can at least call these "direct call to known functions". 

> IUC, the point is that an explicit call to TMC-annotated function foo can be transformed, but when passed as a value, foo "loses" the annotation?

Yes, but this way of explaining things uses a mental model that I think could be more confusing than helpful. (It is a static, first-order program transformation, OCaml functions as values have no notion of "being TMC".) I also don't know how to explain this better so I propose to add "direct", in case it helps some part of the audience, and wait for the examples to clarify this.

dra27 commented on manual/src/refman/extensions/tail_mod_cons.etex:
Ah, maybe that's the terminology - f is a function value or even f is higher order?!?!

gasche replied:
`bind` is higher-order and `f v` is not a direct call / a call to a known function. (I also used "function parameter" which here somehow means both "a parameter that is a function" and the more common "a parameter of the function".)

dra27 commented on manual/src/refman/extensions/tail_mod_cons.etex:
```suggestion
program transformation is.
```
?

ksromanov replied:
This phrase looks to me too weak compared to the rest of the text: it is always better to use tools which you understand.  to rewrite this introductory sentence slightly differently, like

"To discuss the general idea and implementation details of TMC transformation let us recall our 'map' example:"

gasche replied:
> it is always better to use tools which you understand

Yes and no; in general we strive for abstractions that are robust, users don't have to know what is inside to get a good mental model of how to use it. In my experience, TMC is not as good an abstraction, and it helps to know some implementation details to think about its usage. (Not all implementation details matter, but the idea of destination-passing-style certainly does, even though it does not seem apparent in the "interface" of the feature.)

Writing this as a reply to your comment did give me a better idea of something to write:

> To use this advanced feature, it helps to be aware that the function transformation produces a specialized function in destination-passing-style.

Thanks!

dra27 commented on manual/src/refman/extensions/tail_mod_cons.etex:
```suggestion
notation: some operations are not expressible as an OCaml
```

gasche replied:
I went for "not expressible in OCaml source code"

dra27 commented on manual/src/refman/extensions/tail_mod_cons.etex:
```suggestion
something other than a tuple literal will not become tail calls. The
```

dra27 commented on manual/src/refman/extensions/tail_mod_cons.etex:
```suggestion
(x, y, z) = t in f (x, y, z)". If there is any doubt as to whether a call
```

dra27 commented on manual/src/refman/extensions/tail_mod_cons.etex:
```suggestion
can be tail-mod-cons-optimized or not, one can use the "[\@tailcall]"
```

dra27 commented on manual/src/refman/extensions/tail_mod_cons.etex:
```suggestion
Here the "append_flatten" helper is not marked with
```

gasche replied:
The chapter is using both "marked" and "annotated", I guess there was a subtle reason why one or the other felt more natural at the time of writing, but maybe we should use only one form? If so, I think we should go for the "annotated"; it is heavier but some use sites don't feel right with "marked" instead:

> [...] different from the evaluation order of the un-annotated version [...]

> Our recommendation is to use the "[\@tailcall]" annotation on
all callsites that should not consume any stack
space. "[\@tail_mod_cons]" extends the set of functions on which calls
can be annotated to be tail calls [...]

dra27 replied:
Sounds reasonable to me - I think all I was after was the use of a single term (if that was then natural to use everywhere)

dra27 commented on manual/src/refman/extensions/tail_mod_cons.etex:
```suggestion
attribute on the called function, which will warn if the
```

dra27 commented on manual/src/refman/extensions/tail_mod_cons.etex:
```suggestion
programs. Our recommendation is to use the "[\@tailcall]" annotation on
all call sites that should not consume any stack
```

dra27 reviewed (commented):
I'm as always left embarrassed that I can only write documentation in one language, and not necessarily well, even in that one language! Various nits, but most of my comments are optional, so please do disagree.

The terminology at present uses marks, annotations and attributes - I wonder if it might be worth eliminating one of them. The whole thing is implemented with OCaml _attributes_, so perhaps settle on marked or annotated and use that consistently (if you agree, this affects the warning messages in the PR too).

jberdine commented:
Something I think it would help to clarify is what happens if code is "let-expanded" compared to the form shown in examples. For instance what if:
```ocaml
let[@tail_mod_cons] rec map f li =
  match li with
  | [] -> []
  | x :: xs ->
    let y = f x in
    y :: map f xs
```
is instead written:
```ocaml
let[@tail_mod_cons] rec map f li =
  match li with
  | [] -> []
  | x :: xs ->
    let y = f x in
    let ys = map f xs in
    y :: ys
```

johnwhitington commented:
Nitpick:

In some places there are newlines where they aren't needed. Instead of:

```
let[@tail_mod_cons] rec flatten = function
| [] -> []
| xs :: xss ->
  append_flatten xs xss

and[@tail_mod_cons] append_flatten xs xss =
  match xs with
  | [] -> flatten xss
  | x :: xs -> x :: append_flatten xs xss
```

We can have:

```
let[@tail_mod_cons] rec flatten = function
| [] -> []
| xs :: xss -> append_flatten xs xss

and[@tail_mod_cons] append_flatten xs xss =
  match xs with
  | [] -> flatten xss
  | x :: xs -> x :: append_flatten xs xss
```

chambart commented:
While wondering whether there should be references to another chapter telling about tail calls, I realised that there are none. Tail-calls are only mentioned in the attribute section, and that's all. Nothing tells the user that the optimisation exists. I don't think that this should be part of this PR, but I wanted to mention that here.

xavierleroy commented:
On a somewhat related note, I thought that this "chapter" on TMC doesn't really belong as a section in the (already obese) "language extensions" chapter, because TMC is a compilation schema, not a part of the OCaml language.  I was thinking of a new chapter in the "Tools" part, describing high-level aspects of compilation and how they can be controlled with attributes: TMC would be the highlight of this chapter but tail calls certainly deserve a discussion there, maybe poll points too.

gasche commented:
Pierre and I weren't sure about "Language" vs "Tools" as well, I'm happy to move content there. I can add something introductory about tail calls, but I'm definitely not volunteering to cover poll points :-)

ksromanov commented on manual/src/refman/exten.etex:
```diff
@@ -119,4 +119,8 @@ that are implemented in OCaml, but not described in chapter \ref{c:refman}.
 %HEVEA\cutname{bindingops.html}
 \input{bindingops.tex}
 
+\section{s:tail-mod-cons}{Tail Modulo Cons}
```
Is it "Tail Modulo Cons", or "Tail Modulo Constr"? Shouldn't it work not only with `::`?

xavierleroy replied:
I agree the section name should be "Tail-recursion Modulo Constructors".  The "TMC" abbreviation can be introduced and used later.

gasche replied:
done

ksromanov commented on manual/src/refman/extensions/tail_mod_cons.etex:
it is a good place to introduce `TMC` acronym:

```tex
... of tail positions and constructor applications, are \emph{tail modulo
 constructor} (TMC) positions.
 ```

gasche replied:
Good point, done.

ksromanov commented on manual/src/refman/extensions/tail_mod_cons.etex:
In the last line of a snippet we should call `length xs` instead of `length li`. Right?

gasche replied:
Indeed!

ksromanov commented on manual/src/refman/extensions/tail_mod_cons.etex:
Shouldn't it be `on functions` instead of `on programs`? The transformation works on individual functions.

ksromanov commented on manual/src/refman/extensions/tail_mod_cons.etex:
Again, may be "functions" is better here than "programs"?

ksromanov commented on manual/src/refman/extensions/tail_mod_cons.etex:
May be erase "to some functions."?

ksromanov commented on manual/src/refman/extensions/tail_mod_cons.etex:
"For example" was couple lines before. May use "To illustrate" instead?

gasche replied:
I'm told that English readers are not averse to repetition, so I mostly go for consistent usage in technical prose.

ksromanov replied:
Most English readers in our field are not native-English speakers. ;-)

xavierleroy replied:
It's not a question of English vs. other languages; it's a question of technical writing vs. other writing styles.  In technical writing, clarity and precision matter more than elegance.  Using several different words to designate the same thing may look elegant, but it can also be confusing.  So, in technical writing, pick one term for a given concept and use it as many times as necessary.

ksromanov replied:
I am not arguing. It is just a funny thing to point it out.

ksromanov commented on manual/src/refman/extensions/tail_mod_cons.etex:
```tex
\item The transformation is first-order: only direct calls to
  tail-mod-cons functions, known at compile-time, may become tail calls.
  Calls to function values are never optimized with TMC.
  ```

ksromanov commented on manual/src/refman/extensions/tail_mod_cons.etex:
What about "The limitations" -> "These limitations"?

ksromanov commented on manual/src/refman/extensions/tail_mod_cons.etex:
Are "tupled" functions the same as "uncurried" functions?

gasche replied:
Probably so, I'm not sure; the compiler uses "tupled" internally so I went for this naming. ("uncurried" is not used in the manual.)

ksromanov replied:
Thanks!

gadmm commented on manual/src/refman/extensions/tail_mod_cons.etex:
I was in fact expecting an annotation at the callsites when reading the first examples from your documentation. I like this advice. Could you adapt the examples from the documentation so that they follow the documentation's own advice?

gadmm commented on manual/src/refman/extensions/tail_mod_cons.etex:
I am going to criticise this design choice (someone made a comment a bit similar on the original PR—sorry for not writing on the original PR which was a bit technical, it is easier to take a step back at the documentation phase). Whilst the order is still undefined, I noticed that the order is getting more and more consistent towards the right-to-left order over the years. This design is taking a step backwards, which 1) can surprise users, 2) might impede future evolution towards stricter definition of the evaluation order. Have you considered a design where an ambiguity gives a clean rejection of the program, forcing the user to let-expand in this case?

gasche replied:
I don't think `right-to-left` would be the correct evaluation order to specify if we wanted to specify one, obviously left-to-right is the choice that will not "surprise users".

My belief is that pure function calls are common enough that being forced to write programs in A-normal-forms to avoid effect interleaving would feel like a significant *and* unjust burden to users. I made the argument in https://github.com/ocaml/ocaml/pull/9760#issuecomment-727540945 , and my view hasn't really changed since.

If we were to find out that I was totally wrong and this some-argument-goes-last business is a big issue, we could always make things more restrictive in the future (I disagree with your point (2)), as users would then have to rewrite their code in a way that is backward-compatible.

ksromanov replied:
I checked once — almost everywhere the Compiler uses _arabic_ style of evaluation (it is just a matter of fact, **not** the specification):

```ocaml

f arg3 arg2 arg1

[ arg3; arg2; arg1]

arg3::arg2::arg1::[]

( arg3, arg2, arg1 )

[| arg3; arg2; arg1 |]
```

Only 

```ocaml
{|arg3, arg2, arg1|}
```

shows them in the reverse order `arg3, arg2, arg1`. ;-)

--------------------------------------------
If I remember right, this _right-to-left_ order was chosen for efficiency only, and it was first proposed in _Zinc_ - https://caml.inria.fr/pub/papers/xleroy-zinc.pdf (page 14). Nevertheless, Ocaml specification does not fix the evaluation order of the function arguments. Right?

gadmm replied:
Thanks for the pointer to the place this was previously discussed. I think that part of my objection lifts if the annotation on function calls is made mandatory, because the control on evaluation order becomes local. Above I was merely suggesting to apply the advice to the documentation itself, but I also like your idea of making it enforced by the compiler. Obvious subsequent question: if all tail mod const calls were made explicit with an annotation, do you still need to require one on the function definition (edit: assuming an annotation on call sites like `[@tailcall mod_cons]`)? (I did not find a discussion about this point.)

The discussion about changing evaluation order is fairly hypothetical, but I wanted to mention that there are more theoretical arguments in favour of right-to-left evaluation as far as function applications are concerned (I can develop this point elsewhere if you are interested).

gadmm reviewed (commented):
To give an outsider impression of this feature thus described (beyond the amazement and excitement, of course): I kept asking myself throughout the examples of special cases and warnings why there was not just a TRMC annotation on the call sites you wanted to turn TRMC, and a clean error in cases of nonsense/ambiguity (with ambiguities to be resolved by let-expanding). (I am sure that this has been thoroughly discussed/considered, this is just to give an outsider impression of a first contact with the feature. Below I make a related and more specific suggestion, given that you write that the feature is still experimental and can be modified, but this is separate from my outsider feedback.)

AbdallahS commented:
Minor typo in 10.24.4 "(This would the case even if the call to Option.map was inside the Option module.)" -> "would be the case"

gasche committed:
missing line break in TMC warning message

With the previous format, the warnings would print
badly (overflow the line) in the manual.

The underlying problem is that warning messages are passed
around as strings, without formatting hints.

gasche committed:
reformat another TMC warning message

gasche committed:
Apply suggestions from code review

Co-authored-by: wiktorkuchta <35867657+wiktorkuchta@users.noreply.github.com>
Co-authored-by: David Allsopp <david.allsopp@metastack.com>
Co-authored-by: Konstantin Romanov <kromanov1@bloomberg.net>
Co-authored-by: Josh Berdine <josh@berdine.net>
Co-authored-by: John Whitington <john@coherentgraphics.co.uk>

gasche commented:
@gadmm I like your suggestion of always requesting an explicit annotation on TMC calls; it makes the usage of the feature more explicit, which feels good for a first release as we can easily make things less explicit later, but not the other way around. Thanks!

I would like to implement this restriction, but I haven't yet.

gasche committed:
move the tail_mod_cons documentation for "Langage Extensions" to "Tools"

gasche commented:
> On a somewhat related note, I thought that this "chapter" on TMC doesn't really belong as a section in the (already obese) "language extensions" chapter, because TMC is a compilation schema, not a part of the OCaml language. I was thinking of a new chapter in the "Tools" part, describing high-level aspects of compilation and how they can be controlled with attributes: TMC would be the highlight of this chapter but tail calls certainly deserve a discussion there, maybe poll points too.

I moved the documentation from a "language extension" section named "Tail Recursion Modulo Constructors" to a "Tools" chapter named

> The “Tail Modulo Constructor” program transformation

I haven't tried to write more general things about compilation control and/or tail-calls for now.

gasche commented:
At this point I believe that I have addressed all issues that were raised during the review. (Some minor points I did not mark "resolved" in case the original commenter would like to propose an alternative.)

(The blocking `caml_example` issues were swiftly addressed by @Octachron, so things are reasonable on this point. The location chosen to report issues are not great, but there are very few locations available in Lambda where the transformation occurs, and I don't know of an easy way to work around that.)

Two larger tasks remain:
- We could morph the new documentation into something more general also covering tail calls (and poll points?); delayed.
- We could change the implementation to require a `@tailcall` annotation on all transformed tail-modulo-cons callsites. I'm in favor of doing the change before the 4.14 release, but I don't know when I will get around to do it, and I think it should be discussed separately from this documentation PR.

I wonder what remains to be done to consider merging the documentation.

wikku commented on manual/src/cmds/tail-mod-cons.etex:
```diff
@@ -0,0 +1,528 @@
+    Let ((f v, self def), (self[@tailcall]) body)
+\end{caml_example*}
+
+Recommend fix:
```
```suggestion
Recommended fix:
```

wikku commented on manual/src/cmds/tail-mod-cons.etex:
```diff
@@ -0,0 +1,528 @@
+Due to the "[\@tail_mod_cons]" transformation, the calls to "f front"
+and "f rear" will be evaluated before "map f body". In particular,
+this is likely to be different from the evaluation order of the
+un-annotated version. (The evaluation order of constructor arguments
```
```suggestion
unannotated version. (The evaluation order of constructor arguments
```

gasche committed:
simple -> natural?

gasche commented:
@Octachron I think that we probably want the manual chapter in 4.14, if the feature itself is included in the release. This lacks a maintainer-approval, though. I think one could approve on Wiktor's behalf, but doing it myself would be poor form.

Octachron commented on testsuite/tests/tmc/usage_warnings.ml:
```diff
@@ -18,17 +18,19 @@ let[@tail_mod_cons] rec flatten = function
                      ^^^^^^^^^^^^^^^^^^^^^^^
-Warning 72 [tmc-breaks-tailcall]: This call is in tail-modulo-cons position in a TMC function,
+Warning 72 [tmc-breaks-tailcall]: This call
+is in tail-modulo-cons positionin a TMC function,
```
There is typo in the warning itself: `positionin` ⇒ `position in`

amongonz replied:
I've noticed this typo is still present in the 4.14.0 release.

Octachron commented on manual/src/cmds/tail-mod-cons.etex:
```diff
@@ -0,0 +1,528 @@
+adding annotations in the program using attributes, and will be asked
+to do so in any ambiguous situation.
+
+We expect it to be used mostly by advanced OCaml users needing to get
```
This is exclusive `we` seems out of place when the manual (and even this chapter) generally use an inclusive `we`.

Octachron commented on manual/src/cmds/tail-mod-cons.etex:
```diff
@@ -0,0 +1,528 @@
+(Introduced in OCaml 4.14)
+
+Note: this feature is considered experimental, and its interface may
+evolve, with user feedback, in the next few releases of the language.
```
There is a missing introduction here: the reader doesn't know what feature is experimental yet, and the paragraph below starts directly with an explanation of the still mysterious feature. I think it would be better to have a short paragraph or sentence to describe the feature first.

Octachron commented on manual/src/cmds/tail-mod-cons.etex:
```diff
@@ -0,0 +1,528 @@
+  | [] -> []
+  | x :: xs ->
+    let y = f x in
+    let dst = y ::{mutable} Hole in
```
I am not completely sure of how I should read `{mutable} Hole`?

gasche replied:
I guess we could just write `y :: Hole`. My idea with `::{mutable}` was to emphasize that, unlike with normal cons operators, the arguments can be mutated after the fact.

Octachron replied:
I missed the fact that `{mutable}` was a qualifier for `::`. I would propose `::_{mutable}` but that is because I am reading `_` as subscript operator. Maybe  `::[@mutable] Hole`? It is still fantasy syntax but it is clearly an annotation on `::`.

Octachron reviewed (approved):
I have few remarks (and I mostly care about the lack of introduction), but the current chapter looks good overall.

Octachron commented:
@gasche: would you prefer if I merged the chapter and sent my comments as a PR?

gasche commented:
Ah, yes! I had forgotten that this was not merged yet. I think we should get better at merging "text" PRs promptly, because right now everyone has an opinion but no one approves, it's not ideal.

Your plan to merge as-is and improve as a second step sounds good -- better than forgetting about it entirely and not having it in the release, which was our previous plan. (I wouldn't mind helping improve it, but I'm already stretched too thin on OCaml hacking.)

Octachron commented:
Ok, I will merge once either you or me fix the merge conflict.
